File: iOS 8 by Tutorials/Objective-C/Chapter 19 - Beginning Photos/MultipleAssetPicker/MultipleImagePicker/AssetsViewController.m
//
//  AssetCollectionViewController.m
//  Stitch
//
//  Created by Jack Wu on 2014-06-22.
//
//

#import "AssetsViewController.h"
#import "AssetCell.h"

@interface AssetsViewController ()
@property (strong, nonatomic) PHCachingImageManager * imageManager;
@property (strong, nonatomic) NSMutableArray * cachingIndexes;
@property (assign, nonatomic) CGFloat lastCacheFrameCenter;

@property (assign, nonatomic) CGSize assetThumbnailSize;
@property (strong, nonatomic) dispatch_queue_t cacheQueue;

@end

@implementation AssetsViewController

static NSString * const AssetCellReuseIdentifier = @"AssetCell";

#pragma mark - UIViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  self.collectionView.allowsMultipleSelection = YES;
  
  self.imageManager = [[PHCachingImageManager alloc] init];
  self.cachingIndexes = [[NSMutableArray alloc] init];
  self.cacheQueue = dispatch_queue_create("cacheQueue", DISPATCH_QUEUE_SERIAL);
  [self resetCache];
}

- (void)viewWillAppear:(BOOL)animated {
  [super viewWillAppear:animated];
  
  CGFloat scale = [UIScreen mainScreen].scale;
  CGSize cellSize = ((UICollectionViewFlowLayout *)self.collectionViewLayout).itemSize;
  self.assetThumbnailSize = CGSizeMake(cellSize.width * scale, cellSize.height * scale);
  
  [self.collectionView reloadData];
  
  [self updateSelectedItems];
  [self updateCache];
}

- (void)viewWillTransitionToSize:(CGSize)size withTransitionCoordinator:(id<UIViewControllerTransitionCoordinator>)coordinator {
  [super viewWillTransitionToSize:size withTransitionCoordinator:coordinator];
  [self.collectionView reloadData];
  [self updateSelectedItems];
}

- (PHAsset *)currentAssetAtIndex:(NSInteger)index {
  if (self.assetsFetchResults) {
    return self.assetsFetchResults[index];
  }
  else {
    return self.selectedAssets[index];
  }
}

- (void)updateSelectedItems {
  if (self.assetsFetchResults) {
    for (PHAsset * asset in self.selectedAssets) {
      NSUInteger index = [self.assetsFetchResults indexOfObject:asset];
      if (index != NSNotFound) {
        NSIndexPath * indexPath = [NSIndexPath indexPathForItem:index inSection:0];
        [self.collectionView selectItemAtIndexPath:indexPath animated:NO scrollPosition:UICollectionViewScrollPositionNone];
      }
    }
  }
  else {
    //Select all
    for (NSInteger i = 0; i < self.selectedAssets.count; i++) {
      NSIndexPath * indexPath = [NSIndexPath indexPathForItem:i inSection:0];
      [self.collectionView selectItemAtIndexPath:indexPath animated:NO scrollPosition:UICollectionViewScrollPositionNone];
    }
  }
}

#pragma mark - UICollectionViewDelegate

- (void)collectionView:(UICollectionView *)collectionView didDeselectItemAtIndexPath:(NSIndexPath *)indexPath {
  PHAsset * asset = [self currentAssetAtIndex:indexPath.item];
  [self.selectedAssets removeObject:asset];
  if (!self.assetsFetchResults) {
    [self.collectionView deleteItemsAtIndexPaths:@[indexPath]];
  }
}

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
  PHAsset * asset = [self currentAssetAtIndex:indexPath.item];
  [self.selectedAssets addObject:asset];
}

#pragma mark - UICollectionViewDataSource

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
  if (self.assetsFetchResults) {
    return self.assetsFetchResults.count;
  }
  else {
    return self.selectedAssets.count;
  }
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
  AssetCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:AssetCellReuseIdentifier forIndexPath:indexPath];
  
  NSInteger reuseCount = ++cell.reuseCount;
  PHAsset * asset = [self currentAssetAtIndex:indexPath.item];
  
  PHImageRequestOptions * options = [[PHImageRequestOptions alloc] init];
  options.networkAccessAllowed = YES;
  
  [self.imageManager requestImageForAsset:asset
                               targetSize:self.assetThumbnailSize
                              contentMode:PHImageContentModeAspectFill
                                  options:options
                            resultHandler:^(UIImage * result, NSDictionary * info) {
                              if (reuseCount == cell.reuseCount) {
                                cell.imageView.image = result;
                              }
                            }];
  
  return cell;
}

#pragma mark - UICollectionViewDelegateFlowLayout

- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
  NSInteger thumbsPerRow;
  if (collectionView.bounds.size.width < 400) {
    thumbsPerRow = 4;
  }
  else if (collectionView.bounds.size.width < 600) {
    thumbsPerRow = 5;
  }
  else if (collectionView.bounds.size.width < 800) {
    thumbsPerRow = 6;
  }
  else {
    thumbsPerRow = 7;
  }
  
  CGFloat width = collectionView.bounds.size.width / thumbsPerRow;
  CGSize size = {width, width};
  return size;
}

#pragma mark - UIScrollViewDelegate

- (void)scrollViewDidScroll:(UIScrollView *)scrollView {
  dispatch_async(self.cacheQueue, ^{
    [self updateCache];
  });
}

#pragma mark - Caching

- (void)resetCache {
  [self.imageManager stopCachingImagesForAllAssets];
  [self.cachingIndexes removeAllObjects];
  self.lastCacheFrameCenter = 0;
}

- (void)updateCache {
  CGFloat currentFrameCenter = CGRectGetMidY(self.collectionView.bounds);
  if (fabs(currentFrameCenter - self.lastCacheFrameCenter) < CGRectGetHeight(self.collectionView.bounds)/3) {
    // Haven't scrolled far enough yet
    return;
  }
  self.lastCacheFrameCenter = currentFrameCenter;
  
  static NSInteger numOffscreenAssetsToCache = 60;
  
  NSArray * visibleIndexes = [self.collectionView.indexPathsForVisibleItems sortedArrayUsingSelector:@selector(compare:)];
  
  if (!visibleIndexes.count) {
    [self.imageManager stopCachingImagesForAllAssets];
    return;
  }
  
  NSInteger firstItemToCache = ((NSIndexPath *)visibleIndexes[0]).item - numOffscreenAssetsToCache/2;
  firstItemToCache = MAX(firstItemToCache, 0);
  
  NSInteger lastItemToCache = ((NSIndexPath *)[visibleIndexes lastObject]).item + numOffscreenAssetsToCache/2;
  if (self.assetsFetchResults) {
    lastItemToCache = MIN(lastItemToCache, self.assetsFetchResults.count - 1);
  }
  else {
    lastItemToCache = MIN(lastItemToCache, self.selectedAssets.count - 1);
  }
  
  NSMutableArray * indexesToStopCaching = [[NSMutableArray alloc] init];
  NSMutableArray * indexesToStartCaching = [[NSMutableArray alloc] init];
  
  // Stop caching items we scrolled past
  for (NSIndexPath * index in self.cachingIndexes) {
    if (index.item < firstItemToCache || index.item > lastItemToCache) {
      [indexesToStopCaching addObject:index];
    }
  }
  [self.cachingIndexes removeObjectsInArray:indexesToStopCaching];
  
  [self.imageManager stopCachingImagesForAssets:[self assetsAtIndexPaths:indexesToStopCaching]
                                     targetSize:self.assetThumbnailSize
                                    contentMode:PHImageContentModeAspectFill
                                        options:nil];
  
  // Start caching new items in range
  for (NSInteger i = firstItemToCache; i < lastItemToCache; i++) {
    NSIndexPath * index = [NSIndexPath indexPathForItem:i inSection:0];
    if (![self.cachingIndexes containsObject:index]) {
      [indexesToStartCaching addObject:index];
      [self.cachingIndexes addObject:index];
    }
  }
  
  [self.imageManager startCachingImagesForAssets:[self assetsAtIndexPaths:indexesToStartCaching]
                                      targetSize:self.assetThumbnailSize
                                     contentMode:PHImageContentModeAspectFill
                                         options:nil];
  
}

- (NSArray *)assetsAtIndexPaths:(NSArray *)indexPaths {
  if (!indexPaths.count) {
    return nil;
  }
  
  NSMutableArray *assets = [NSMutableArray arrayWithCapacity:indexPaths.count];
  
  for (NSIndexPath *indexPath in indexPaths) {
    PHAsset *asset = [self currentAssetAtIndex:indexPath.item];
    [assets addObject:asset];
  }
  
  return assets;
}

@end

File: iOS 8 by Tutorials/Swift/Chapter 19 - Beginning Photos/MultipleAssetPicker-Starter/MultipleAssetPicker/AssetsViewController.swift
/*
* Copyright (c) 2014 Razeware LLC
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
*/

import UIKit
import Photos

class AssetsViewController: UICollectionViewController, UICollectionViewDelegateFlowLayout {
  let AssetCollectionViewCellReuseIdentifier = "AssetCell"
  
  var assetsFetchResults: PHFetchResult?
  var selectedAssets: SelectedAssets!
  
  private var assetThumbnailSize = CGSizeZero
  
  // MARK: UIViewController
  override func viewDidLoad() {
    super.viewDidLoad()
    collectionView.allowsMultipleSelection = true
  }

  override func viewWillAppear(animated: Bool)  {
    super.viewWillAppear(animated)
    
    // Calculate Thumbnail Size
    let scale = UIScreen.mainScreen().scale
    let cellSize = (collectionViewLayout as UICollectionViewFlowLayout).itemSize
    assetThumbnailSize = CGSize(width: cellSize.width * scale, height: cellSize.height * scale)
    
    collectionView.reloadData()
    updateSelectedItems()
  }
  
  override func viewWillTransitionToSize(size: CGSize, withTransitionCoordinator coordinator: UIViewControllerTransitionCoordinator) {
    super.viewWillTransitionToSize(size, withTransitionCoordinator: coordinator)
    collectionView.reloadData()
    updateSelectedItems()
  }
  
  // MARK: Private
  func currentAssetAtIndex(index:NSInteger) -> PHAsset {
    if let fetchResult = assetsFetchResults {
      return fetchResult[index] as PHAsset
    } else {
      return selectedAssets.assets[index]
    }
  }
  
  func updateSelectedItems() {
    // Select the selected items
  }
  
  // MARK: UICollectionViewDelegate
  override func collectionView(collectionView: UICollectionView, didSelectItemAtIndexPath indexPath: NSIndexPath)  {
    // Update selected Assets
  }
  
  override func collectionView(collectionView: UICollectionView, didDeselectItemAtIndexPath indexPath: NSIndexPath)  {
    // Update selected Assets
  }
  
  
  // MARK: UICollectionViewDataSource
  override func collectionView(collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
    if let fetchResult = assetsFetchResults {
      return fetchResult.count
    } else if selectedAssets != nil {
      return selectedAssets.assets.count
    }
    return 0
  }
  
  override func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
    let cell = collectionView.dequeueReusableCellWithReuseIdentifier(AssetCollectionViewCellReuseIdentifier, forIndexPath: indexPath) as AssetCell
    
    // Populate Cell
    
    return cell
  }
  
  // MARK: UICollectionViewDelegateFlowLayout
  func collectionView(collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAtIndexPath indexPath: NSIndexPath) -> CGSize {
    var thumbsPerRow: Int
    switch collectionView.bounds.size.width {
    case 0..<400:
      thumbsPerRow = 4
    case 400..<600:
      thumbsPerRow = 5
    case 600..<800:
      thumbsPerRow = 6
    case 800..<1200:
      thumbsPerRow = 7
    default:
      thumbsPerRow = 4
    }
    let width = collectionView.bounds.size.width / CGFloat(thumbsPerRow)
    return CGSize(width: width,height: width)
  }
}
